libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+            }
+            context.error = undefined;
+          }
+          const { cacheId, contextId } = cobuildLock.cobuildContext;
+
+          const finalCacheId: string =
+            status === OperationStatus.Failure ? `${cacheId}-${contextId}-failed` : cacheId;
+          switch (status) {
+            case OperationStatus.SuccessWithWarning:
+            case OperationStatus.Success:
+            case OperationStatus.Failure: {
+              const currentStatus: ICobuildCompletedState['status'] = status;
+              setCompletedStatePromiseFunction = () => {
+                return cobuildLock?.setCompletedStateAsync({
+                  status: currentStatus,
+                  cacheId: finalCacheId
+                });
+              };
+              setCacheEntryPromise = cobuildLock.projectBuildCache.trySetCacheEntryAsync(
+                buildCacheTerminal,
+                finalCacheId
+              );
+            }
+          }
+        }
+
+        // If the command is successful, we can calculate project hash, and no dependencies were skipped,
+        // write a new cache entry.
+        if (!setCacheEntryPromise && taskIsSuccessful && isCacheWriteAllowed && projectBuildCache) {
+          setCacheEntryPromise = projectBuildCache.trySetCacheEntryAsync(buildCacheTerminal);
+        }
+        const cacheWriteSuccess: boolean | undefined = await setCacheEntryPromise;
+        await setCompletedStatePromiseFunction?.();
+
+        if (cacheWriteSuccess === false && afterExecuteContext.status === OperationStatus.Success) {
+          afterExecuteContext.status = OperationStatus.SuccessWithWarning;
+        }
+
+        return afterExecuteContext;
+      }
+    );
+  }
+
+  private _getBuildCacheContextByRunner(runner: IOperationRunner): IOperationBuildCacheContext | undefined {
+    const buildCacheContext: IOperationBuildCacheContext | undefined =
+      this._buildCacheContextByOperationRunner.get(runner);
+    return buildCacheContext;
+  }
+
+  private _getBuildCacheContextByRunnerOrThrow(runner: IOperationRunner): IOperationBuildCacheContext {
+    const buildCacheContext: IOperationBuildCacheContext | undefined =
+      this._getBuildCacheContextByRunner(runner);
+    if (!buildCacheContext) {
+      // This should not happen
+      throw new InternalError(`Build cache context for runner ${runner.name} should be defined`);
+    }
+    return buildCacheContext;
+  }
+
+  private async _tryGetProjectBuildEnabledAsync({
+    buildCacheConfiguration,
+    rushProject,
+    commandName
+  }: {
+    buildCacheConfiguration: BuildCacheConfiguration;
+    rushProject: RushConfigurationProject;
+    commandName: string;
+  }): Promise<boolean> {
+    const consoleTerminalProvider: ConsoleTerminalProvider = new ConsoleTerminalProvider();
+    const terminal: ITerminal = new Terminal(consoleTerminalProvider);
+    // This is a silent terminal
+    terminal.unregisterProvider(consoleTerminalProvider);
```

Comment:
I've created NullTerminalProvider equivalents before, we should just have a utility for this.

That said, this will get cleaned up when refactoring `RushProjectConfiguration` instances to be bulk read prior to operation graph creation, which will be necessary for other feature asks, since that is the only thing the terminal is being used for.

Reply:
Hi David, I don't notice there is a `NullTerminalProvider` in the current codebase. Could you kindly point me out where is the code. I would like to use this utility.